tests/integration/coordinator/plugins/clickhouse/test_plugin.py

```diff
@@ -343,6 +377,50 @@ async def test_restores_materialized_view_data(restored_cluster: Sequence[ClickH
         assert response == expected_data
 
 
+async def test_restores_materialized_view_deleted_source_table(restored_cluster: Sequence[ClickHouseClient]) -> None:
+    s1_data = [[7 * 3]]
+    s2_data = [[9 * 3]]
+    cluster_data = [s1_data, s1_data, s2_data]
+    for client, expected_data in zip(restored_cluster, cluster_data):
+        response = await client.execute(b"SELECT thekey3 FROM default.materialized_view_deleted_source ORDER BY thekey3")
+        assert response == expected_data
+    client = restored_cluster[0]
+    # Recreated deleted table works again
+    await client.execute(
+        b"CREATE TABLE default.source_table_for_materialized_view_deleted (thekey UInt32, thedata String) "
+        b"ENGINE = ReplicatedMergeTree ORDER BY (thekey)"
+    )
+    await client.execute(b"INSERT INTO default.source_table_for_materialized_view_deleted VALUES (8, '8')")
+    s1_data = [[7 * 3], [8 * 3]]
+    s2_data = [[9 * 3]]
+    cluster_data = [s1_data, s1_data, s2_data]
+    for client, expected_data in zip(restored_cluster, cluster_data):
+        response = await client.execute(b"SELECT thekey3 FROM default.materialized_view_deleted_source ORDER BY thekey3")
+        assert response == expected_data
+
+
+async def test_restores_view_deleted_source_table(restored_cluster: Sequence[ClickHouseClient]) -> None:
+    unknown_table_exception_code = 60
+    for client in restored_cluster:
+        with pytest.raises(ClickHouseClientQueryError) as raised:
+            await client.execute(b"SELECT thekey2 FROM default.view_deleted_source ORDER BY thekey2")
+        assert raised.value.status_code == 404
+        assert raised.value.exception_code == unknown_table_exception_code
+    # Recreated deleted table works again
+    await restored_cluster[0].execute(
+        b"CREATE TABLE default.source_table_for_view_deleted (thekey UInt32, thedata String) "
+        b"ENGINE = ReplicatedMergeTree ORDER BY (thekey)"
+    )
+    await restored_cluster[0].execute(b"INSERT INTO default.source_table_for_view_deleted VALUES (11, '11')")
+    await restored_cluster[2].execute(b"INSERT INTO default.source_table_for_view_deleted VALUES (17, '17')")
```

Comment:
Sorry I didn't notice earlier.

This test will mutate the `restored_cluster` that has a module scope, this will affect other tests.

This needs a separate `function_restored_cluster` fixture which does the same as the module fixture but function-scoped and without the steps parametrization. (We don't want to make all tests function scoped, since the restore is slow and this test doesn't need to check the partial-restore-and-retry that we use for other tests)

Reply:
Sorry, I did it intentionally, I thought it is fine. But probably it is better to not touch state which looks immutable.
The fix looks a bit ugly, but fine. I am not exactly sure what this test is testing, but let it better be there, as someone will want to `undelete` some table one day.